### PR TITLE
impl `grant_incentive_bonus` for bridge

### PR DESCRIPTION
**Pull request by rjman-ljm**
Feature:
  + Add gift logic for `bridge.materialize`
Close:
  + #1403

### CODE REVIEW

File: .github/workflows/dockerimage.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ name: Docker Image
 
 on:
   push:
-    branches: [master]
+    branches: [master, gift-fee]
```
Comment (rjman-ljm) [inline .github/workflows/dockerimage.yml:5]: will remove it later

File: primitives/src/currency.rs [hunk 2]
```diff
@@ -174,3 +180,9 @@ impl<
         Ok(asset.clone().into())
     }
 }
+pub struct MockGiftConvert;
```
Comment (GopherJ) [inline primitives/src/currency.rs:183]: Mock code shouldn't be here

File: primitives/src/lib.rs [hunk 3]
```diff
@@ -99,6 +99,8 @@ pub type PriceDetail = (Price, Timestamp);
 
 pub type TimeStampedPrice = orml_oracle::TimestampedValue<Price, Moment>;
 
+pub type CurrencyDetail = (CurrencyId, Balance);
```
Comment (GopherJ) [inline primitives/src/lib.rs:102]: feel we dont need this type and `CurrencyDetail` cannot describe amount

File: runtime/heiko/src/lib.rs [hunk 4]
```diff
@@ -925,16 +925,26 @@ parameter_types! {
 }
 
 pub struct GiftConvert;
-impl Convert<Balance, Balance> for GiftConvert {
-    fn convert(amount: Balance) -> Balance {
-        let decimal = <Assets as InspectMetadata<AccountId>>::decimals(&KSM);
+impl Convert<CurrencyDetail, Balance> for GiftConvert {
+    fn convert(currency: CurrencyDetail) -> Balance {
+        let decimal = <Assets as InspectMetadata<AccountId>>::decimals(&currency.0);
```
Comment (GopherJ) [inline runtime/heiko/src/lib.rs:930]: it's preferred to expand instead of using `currency.0`

File: runtime/heiko/src/lib.rs [hunk 5]
```diff
@@ -925,16 +925,26 @@ parameter_types! {
...
-        if amount >= 10_u128.pow((decimal - 1).into()) {
-            return DOLLARS / 40; // 0.025HKO
+        match currency {
+            (EUSDT | EUSDC, amount) => {
+                // greater than 300 EUSDT/EUSDC
+                if amount >= 300 * 10_u128.pow(decimal.into()) {
+                    return DOLLARS / 40; // 0.025HKO
+                }
+            }
+            (_, amount) => {
```
Comment (GopherJ) [inline runtime/heiko/src/lib.rs:942]: this should work like a whitelist so by default we dont give any token

File: runtime/kerria/src/lib.rs [hunk 6]
```diff
@@ -933,22 +933,31 @@ parameter_types! {
 }
 
 pub struct GiftConvert;
-impl Convert<Balance, Balance> for GiftConvert {
-    fn convert(amount: Balance) -> Balance {
-        let decimal = <Assets as InspectMetadata<AccountId>>::decimals(&DOT);
+impl Convert<CurrencyDetail, Balance> for GiftConvert {
+    fn convert(currency: CurrencyDetail) -> Balance {
+        let decimal = <Assets as InspectMetadata<AccountId>>::decimals(&currency.0);
```
Comment (GopherJ) [inline runtime/kerria/src/lib.rs:938]: same like above

File: rust-toolchain [hunk 7]
```diff
@@ -1 +1 @@
-nightly-2021-11-07
+nightly
```
Comment (GopherJ) [inline rust-toolchain:1]: this better to be a nightly version which is stable enough for debugging (without too much ICE)

Comment (GopherJ) [reply rust-toolchain:1]: and `nightly-2021-11-07` is also used in CI, if we would like to change, we should pick another better version then replace in all places

Comment (rjman-ljm) [reply rust-toolchain:1]: sorry, I just want to change something to commit and trigger CI, I will change this back now.

File: pallets/bridge/src/lib.rs [hunk 8]
```diff
@@ -810,10 +826,33 @@ impl<T: Config> Pallet<T> {
...
+    #[require_transactional]
     fn cancel_materialize(src_id: ChainId, src_nonce: ChainNonce) -> DispatchResult {
         Self::update_bridge_registry(src_id, src_nonce);
         Self::deposit_event(Event::ProposalRejected(src_id, src_nonce));
 
         Ok(())
     }
+
+    #[require_transactional]
+    fn gift_fees(who: T::AccountId, asset_id: CurrencyId, amount: BalanceOf<T>) -> DispatchResult {
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:838]: gift_fees looks like a getter but there are mutations inside

File: pallets/bridge/src/mock.rs [hunk 9]
```diff
@@ -171,6 +174,36 @@ parameter_types! {
...
+            HKO => {
+                // greater than 0.1 Token
+                if balance >= 10_u128.pow((decimal - 1).into()) {
+                    return Ok(default_gift_amount);
+                }
+            }
+            _ => return Ok(Zero::zero()),
+        }
+
+        Ok(Zero::zero())
```
Comment (GopherJ) [inline pallets/bridge/src/mock.rs:205]: this line can be removed

File: pallets/bridge/src/mock.rs [hunk 10]
```diff
@@ -171,6 +174,36 @@ parameter_types! {
...
+        }
+        let default_gift_amount = 1_000_000_000_000 / 40; // 0.025HKO
+        match asset_id {
+            USDT => {
+                // greater than 300 EUSDT/EUSDC
+                if balance >= 300 * 10_u128.pow(decimal.into()) {
+                    return Ok(default_gift_amount);
+                }
+            }
+            HKO => {
```
Comment (GopherJ) [inline pallets/bridge/src/mock.rs:196]: can be merged `HKO if ... {}`

File: pallets/bridge/src/lib.rs [hunk 11]
```diff
@@ -810,10 +826,33 @@ impl<T: Config> Pallet<T> {
...
+        let beneficiary_native_balance =
+            T::Assets::reducible_balance(native_currency_id, &who, true);
+        let reducible_balance =
+            T::Assets::reducible_balance(native_currency_id, &gift_account, false);
+
+        if !gift_amount.is_zero()
+            && reducible_balance >= gift_amount
+            && beneficiary_native_balance < gift_amount
+        {
+            let diff = T::ExistentialDeposit::get() + gift_amount - beneficiary_native_balance;
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:852]: this ExistentialDeposit is only for native balance

File: runtime/kerria/src/lib.rs [hunk 12]
```diff
@@ -933,19 +934,31 @@ parameter_types! {
...
+                // greater than 300 EUSDT/EUSDC
+                if balance >= 300 * 10_u128.pow(decimal.into()) {
+                    return Ok(default_gift_amount);
+                }
+            }
+            _ => return Ok(Zero::zero()),
         }
 
-        Zero::zero()
+        Ok(Zero::zero())
```
Comment (GopherJ) [inline runtime/kerria/src/lib.rs:961]: same like above

File: runtime/kerria/src/lib.rs [hunk 13]
```diff
@@ -933,19 +934,31 @@ parameter_types! {
...
-            return Zero::zero();
+            return Ok(Zero::zero());
         }
-
-        // 5DOT
-        if amount >= 5 * 10_u128.pow(decimal.into()) {
-            return 125 * DOLLARS / 100; // 1.25PARA
+        let default_gift_amount = 125 * DOLLARS / 100; // 1.25PARA
+        match asset_id {
+            NATIVE_ASSET_ID => {
```
Comment (GopherJ) [inline runtime/kerria/src/lib.rs:946]: same like above

**Comment by GopherJ** [comment]
LGTM except the comments

**Review by GopherJ — APPROVED** [review]

File: pallets/bridge/src/lib.rs (hunk 8 above)
Comment (rjman-ljm) [reply pallets/bridge/src/lib.rs:838]: > gift_fees looks like a getter but there are mutations inside

how is `grant_incentive_bonus`?

Comment (GopherJ) [reply pallets/bridge/src/lib.rs]: yeah it works

**Pull request merged**